tests/python/pants_test/pantsd/test_pailgun_server.py

```diff
@@ -78,6 +82,102 @@ def test_process_request_thread_error(self, mock_shutdown_request):
     self.assertIs(self.mock_handler_inst.handle_error.called, True)
     mock_shutdown_request.assert_called_once_with(self.server, mock_request)
 
+  def test_ensure_request_is_exclusive(self):
+    """Launch many requests, assert that every one is trying to enter the critical section, and assert that only one is doing so at a time."""
+    self.threads_to_start = 10
+
+    self.threads_running_cond = threading.Condition()
+    self.threads_running = 0
+    def handle_thread_tried_to_handle_request():
```

Comment:
Maybe call this `handle_thread_started()`?

Reply:
I like the overly descriptive names for tests. In this context, `started` can mean a number of things, and indeed a thread can be started but have not reached this function. What we want to assert here is that a thread has arrived to the beginning of the critical section.